subprojects/language-java/src/main/java/org/gradle/api/tasks/javadoc/Javadoc.java

```diff
@@ -146,6 +147,19 @@ protected void generate() {
         options.classpath(new ArrayList<>(javaModuleDetector.inferClasspath(isModule, getClasspath()).getFiles()));
         options.modulePath(new ArrayList<>(javaModuleDetector.inferModulePath(isModule, getClasspath()).getFiles()));
 
+        // If modularized JavaDoc is needed, we need to add the source paths - the file listing is not enough alone
+        // See #19726 for more
+        if (isModule) {
```

Comment:
Is it possible to allow non-module users to opt into this behavior, or make it the default?

https://github.com/gradle/gradle/issues/19869 shows a case where the javadoc task breaks depending on what dependencies have been declared.

Reply:
The `--source-path` is also introduced in [MinimalJavadocOption](https://github.com/gradle/gradle/pull/19891/files#diff-a0fa9a560c40655c0d3b841158d8f52b44e3a0c0a6ecd9d9c526ae210adc113fR338), so there should be no issue with using it alone.  
This code relies on the same external API as you would from the script.

So if you use the `options.sourcePath`, it should appear in the `javadoc.options` file, and you could use it as you wish.
You meant something like this?

Reply:
IMO, setting `--source-path` with the known source directories should be the default behavior of the Javadoc task. Are there use cases that you expect will break by making it a default? Or, failing tests? For use cases that depend on the source-path == classpath, it should be easy for them to opt-in by unsetting source-path.

I do see the value in exposing source-path in MinimalJavadocOptions; but if the current logic only sets it for `isModule`, I expect the majority of setters will be duplicating logic to achieve the same ends that could have otherwise been the default.

Reply:
First of all, JavaDoc' s documentation is... having something to be desired, as we looked at it for a long time, and had to trial-and error a lot even after reading it letter by letter.

> setting --source-path with the known source directories should be the default behavior of the Javadoc task. 

This is an elegant solution, but sadly it seems that we cannot use it alone. When you have exclusions, you need to communicate with JavaDoc. What we found (in the modularized case), is that we can only give JavaDoc a partial file list, if we supply **all** the source files. 

>  do see the value in exposing source-path in MinimalJavadocOptions; but if the current logic only sets it for isModule, I expect the majority of setters will be duplicating logic to achieve the same ends that could have otherwise been the default.

This part of the code is the workaround to make modularized+excluded projects work. As I said, the `options.sourcePath(...)` is not limited by this logic, that is a new option, which can be used anytime by anybody.


Apart from this entire bug solving, I would be glad if you know something why applying `--source-path` every time would be beneficial. With the state of the JavaDoc documentation, and the sensitivity of the task (it's used by a lot of people), I am feeling uneasy making big changes, as I'm afraid the modification would break a lot of people's build. Yes, we have a lot of tests, but the world is big, and again, I'm trying to limit the possible blast-radius of this modification.

Reply:
I didn't mean to imply that setters wouldn't be able to use `options.sourcePath(...)`; I meant to imply that they will be replicating the logic to create the proper sourcepath value.

I'm *not* a Javadoc expert, just making observations based on #19869 where depending on the jar [com.google.jsinterop:base:1.0.0](https://search.maven.org/artifact/com.google.jsinterop/base/1.0.0/jar) breaks the Javadoc task. For my very small use case, it would be nice if it were a Javadoc Task default; but I completely understand the desire to limit the scope of potential breaking changes.

Thanks for this PR, I think it will be *great* to have `options.sourcePath(...)`. I'll update #19869 and corresponding repo with `options.sourcePath(...)` after it hits the next release.

Reply:
Thank _you_ for providing useful info with this!
I think it's the best to keep this PR moving, and consider further ideas in #19869.